Enable user control of process pinning in launcher

cmcantalupo opened the pull request:
This adds a new command line option to geopmlaunch --geopm-affinity-disable which will allow the user to control all cpu affinity options for the underlying launch command.  This is coupled with a change to the ApplicationSampler::connect() method that affinitizes the thread using the sampler to a core not used by the application.

cmcantalupo committed:
Pin application sampler process at end of connect() method

- Pins the thread using the ApplicationSampler (i.e. the controller)
  to the highest linux logical cpu not used by the application, or
  linux logical cpu 0 if all CPUs are used by application.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add feature to launcher to disable process pinning

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on src/ApplicationSampler.cpp:
```diff
@@ -371,8 +372,51 @@ namespace geopm
+    int ApplicationSamplerImp::sampler_cpu(void)
+    {
+        int result = 0;
+        for (int cpu_idx = m_num_cpu - 1; cpu_idx != -1; --cpu_idx) {
```
Could land the controller on hyperthread of application if it is using the last core, but not the hyperthread of the last core.

cmcantalupo commented on src/ApplicationSampler.cpp:
```diff
@@ -371,8 +372,51 @@ namespace geopm
             }
         }
+        // Try to pin the sampling thread to a free core
+        std::vector<bool> cpu_enabled(m_num_cpu, false);
```
cpu_enabled -> sampler_cpu_set

cmcantalupo committed:
Run GEOPM on an isolated core, not just an isolated CPU

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up implementation of ApplicationSampler affinity

- Add unit tests to cover new code.
- Clean up messy return type of make_cpu_set() with auto
- Move cpu_set_t unique_ptr implementation into separate file

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update the README to reflect changes in affinity implementation

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update geomplaunch man page with --geopm-affinity-disable doc

- Also add help text to geopmlaunch --help

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on README:
```diff
@@ -375,15 +375,16 @@ script.
+geopmlaunch(1)).
+
+While the GEOPM control thread connects to the application it will
+automatically affinitize iteself to the highest indexed core not used
```
```suggestion
automatically affinitize itself to the highest indexed core not used
```

bgeltz commented on src/ApplicationSampler.cpp:
```diff
@@ -371,6 +373,34 @@ namespace geopm
+        // Try to pin the sampling thread to a free core
+        std::set<int> sampler_cpu_set = {sampler_cpu()};
+        auto sampler_cpu_mask = make_cpu_set(m_num_cpu, sampler_cpu_set);
+        (void)sched_setaffinity(0, CPU_ALLOC_SIZE(m_num_cpu), sampler_cpu_mask.get());
```
Don't ignore the rc.

bgeltz commented on test/ApplicationSamplerTest.cpp:
```diff
@@ -525,3 +547,25 @@ TEST_F(ApplicationSamplerTest, cpu_progress)
 }
+
+TEST_F(ApplicationSamplerTest, sampler_cpu)
+{
```
Comments please to explain the numbering here between physical cores and linux logical cpus.

bgeltz committed:
Update README

Signed-off-by: Brad Geltz <brad.geltz@intel.com>

cmcantalupo committed:
Add warning if setting affinity of sampling thread fails

- Also add some comments to ApplicationSamplerTest.sampler_cpu to
  clarify what is going on.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add clarifying parens

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Do not allocate an extra core for the process mask with pthread launch

- Now that the controller is migrated to a CPU outside of the
  process masks, we do not want to include space for it.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix affinity tests to reflect change in mask for pthread launch mode

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
